### ISSUE TITLE
haxe.Template parenthesis

### ISSUE THREAD

**Issue by filt3rek** [HaxeFoundation/haxe#10202]
Hej,

I'm sorry to come back with that, but I don't know if it's a specification that a condition must be between 2 expressions and not more, so you have to group conditions in parenthesis in order to have always 2 expressions, or if it's a bug.
I found that not very cool, it doesn't "match" haxe language even if I know that template is not haxe...

Please take a look at that : https://try.haxe.org/#32C73168
```haxe
class Test {
  static function main() {
    //var s = "Blabla::if((a==1)||(a==2))::Foo::end::"; // OK
    //var s = "Blabla::if((a==1)||(a==2)||(a==3))::Foo::end::"; // Fail
    //var s = "Blabla::if((a==1)||((a==2)||(a==3)))::Foo::end::"; // Ok
    var s = "Blabla::if(((a==1)||(a==2))||((a==3)||(a==4)))::Foo::end::"; // Ok
    var tpl	= new haxe.Template( s );
    var out	= tpl.execute( { a : 1 } );
    trace( out );
  }
}
```

Thanks for your attention,
Regards.

**Comment by back2dos** [comment]
Binary operations are specified to require parentheses: https://haxe.org/manual/std-template.html#expressions

**Comment by filt3rek** [comment]
> Binary operations are specified to require parentheses: https://haxe.org/manual/std-template.html#expressions

@back2dos : Did you read my issue ?...

**Comment by kLabz** [comment]
I suppose Juraj was answering that part:
> but I don't know if it's a specification that a condition must be between 2 expressions

**Comment by filt3rek** [comment]
Yes but maybe I was not clear especially in my title, the problem is not on parenthesis but the fact that there must be at most 2 expressions, that why when you have 3 expressions or more, you have to put parenthesis that way to reduce expressions to only 2.
Hehe, I don't know if this explanation is more clear :rofl:

**Comment by kLabz** [comment]
Oh, yeah sorry, his message wasn't explicitely talking about the 2 expressions requirement but the link mentions this:

> Operators can be used but they don't deal with operator precedence. Therefore it is required to enclose each operation in parentheses ().

In itself it doesn't forbid things like `(1 + 2 + 3)` (or does it? xD), but I checked the code and operations do require exactly 2 expressions (see https://github.com/HaxeFoundation/haxe/blob/4.2.1/std/haxe/Template.hx#L371), and I'm not sure supporting more would be an easy change.

**Comment by back2dos** [comment]
A binary operation is always `(a op b)` (as opposed to `a op b`). This includes `||`, meaning that `e1 || e2` is invalid syntax and the correct form is `(e1 || e2)`. Also note that `if` does not require parentheses, so in `if((a==1)||(a==2))`, the expression is `((a==1) || (a==2))`, which is of the form `(e1 || e2)` and therefore parses just fine.

This means that `((a==1)||(a==2)||(a==3))` will not parse, because `(e1 op e2 op e3)` is not valid syntax (instead you need `((e1 op e2) op e3)` or `(e1 op (e2 op e3))`). This is not a bug, but a result of `haxe.Template`'s simplistic syntax. Always requiring parentheses around binary operations saves one the trouble of dealing with operator precedence and associativity.

So this is not a bug report, but a feature request.

**Comment by filt3rek** [comment]
Ok guys, as I wrote at the begining, I don't know if it's a bug or a `haxe.Template` specification and at the end I also wrote that 
> I found that not very cool, it doesn't "match" haxe language even if I know that template is not haxe...

So it was also a kind of suggestion to fit that more friendly. (Yes Juraj, a kind of feature request, all was written in my first comment :))

Seeing Aleksandr taking that as a bug, I thought it will be improved.

But if you think It has to be left as is, ok.
So if this "issue" will be closed, just maybe that should be more explicitly written in the specs, I don't know, just another suggestion :)

That's all for me, have nice day !
